### ISSUE TITLE
P.S. Happy new year

### ISSUE THREAD

**Issue by yartem** [neuromation/cookiecutter-neuro-project#320]
Makefile+cookiecutter was adopted as a temporary proof-of-concept project aimed to show how much we can simplify work with `neuro` with a higher-level abstraction. This decision was a great finding as a temporary solution, but now it grew up to a solution that does not have any advantage (at least from my point of view).

Here's arguments for using Makefile+cookiecutter and my thoughts to them in sub-items:

1. We assume that the users know syntax of Makefile and thus they can take a look into it to see how it works.
1.1 Please take a look into [current Makefile](https://github.com/neuromation/cookiecutter-neuro-project/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/Makefile) or especially to the [hyper-parameter search implementation](https://github.com/neuromation/cookiecutter-neuro-project/pull/307/files#diff-21cd6e8e5fd83b2e927adbe64c6966e6R338-R382) and tell me that this code is readable and maintainable).
2. Makefile doesn't work on Windows
2.1 Yes, technically one can install it via cygwin or something (have you tried?). But in order to have some basic logic (see for example the same [hyper-parameter search implementation](https://github.com/neuromation/cookiecutter-neuro-project/pull/307/files#diff-21cd6e8e5fd83b2e927adbe64c6966e6R338-R382)) we need to use `grep`, `tee`, `tail`, `rev` -- are they available on Windows?
3. Updating the project. 
3.1 We don't separate targets written by us from custom user-defined targets.
3.2 So technically, we put the burden of writing Makefile targets on the user -- at the moment when they will be resolving git conflicts while upgrading their project, they will leave our platform.
3.3 At least what we should do is split Makefile into several ones to hide that over-over-overcomplicated code ("black magic code" according to @mariyadavydova) so that the users don't have conflicts.
4. Makefile was a quick-to-implement solution.
4.1 Ha-ha, not at all. I spent from 5 to 10 times more time on struggling with dumb `make` problems and limitations comparing to what I would have spent with Python.
4.2 As a result -- much poorer functionality rather than it could have been written on Python.

I see the good solution to be based on neuro aliases, which @asvetlov have been working on recently:
1. `neuro` has a command group `project`, which works only in a specific directory similarly to git (most of git commands don't work if there's no `.git/` directory)
2. once the project have been initialized via cookiecutter (`neuro project init`), the user goes to the project directory and gets the `neuro project` commands unlocked:
3. `neuro project upload-data`, `neuro project jupyter` -- all the commands available on Makefile can be available with `neuro project`.
4. To avoid writing `neuro project` on every command, we shorten them with neuro aliases: `neuro upload-data`, `neuro jupyter`.

Technically, this can be implemented only on top of neuro aliases only if they support arguments. 

@asvetlov if you agree on this concept, I would love to work on designing neuro project abstraction on top of argumentful neuro aliases starting from January 20, 2020.

cc @astaff @mariyadavydova @dalazx.

P.S. Happy new year.

**Comment by asvetlov** [comment]
Please don't rush. 
Aliases don't exist yet; even when they exist I doubt if pushing complex multi-line code into an alias is a good idea.

**Comment by mariyadavydova** [comment]
@ayushkovskiy Thank you for putting your thoughts on this subject together. This is indeed a complex question, and I would like to comment come of your points.

1. Users and Makefile: I agree that the complexity of some of the targets is way more than the others (compare uploading code to hyper-parameter tuning). I also agree that our customers are probably able to read Python code. However, I'm not sure if hyper-parameter tuning implementation inside a Python script is much more readable and maintainable.

2. Makefile on Windows: This is a very valid point. We rely on Unix tools, which won't work on Windows. At the same time, our stats show that 37% of the website visitors in December come from Windows (with 25% from OS X, 12% from Linux, and everything else from mobile or other platforms). We may read this as "Half of our potential users can't use some parts of our template". 

3. Upgrading the project: Yes, there is a problem, and I still don't understand why we can't solve it with our Makefile. However, I have an assumption that this problem is an internal one. We need the template to be easily upgraded to keep our recipes up-to-date. Our customers will hardly think about it, because we keep the CLI backwards compatible, and the template nature is to be a starting point, not a framework. This assumption can only be verified over the time.

4. Makefile development: I understand your pain here. A Makefile is not really simple to develop when it comes to something more complex than "Several constants and several one-line targets". However, will you disagree that the result is pretty simple to use, at least when starting the work on a project?

**Comment by mariyadavydova** [comment]
Now, speaking of possible solutions. I assume that we should keep our project template features in mind. I would say that the template should be:

1. Flexible (a user may easily add their tools in the workflow or change the current tools behaviour).
2. Easy to read (an average ML engineer can understand the code).
3. Cross-platform (Linux, OS X, Windows).
4. Easy to develop (for us).
5. Upgradable (to keep the recipes up-to-date).

I put these features in order of "Most important -> least important", meaning that the comfort of our users (1, 2, 3) is more important than our own comfort (4, 5).

Now, we can evaluate each possible solution against these features. I would prefer to do that accurately later, after vacations. 

In short, 
* Makefile is very flexible, more or less easy to read, not cross-platform, hard to develop, and more or less upgradable.
* Alias-based solution seems to be way less flexible. I'm not even sure if one can support all the same functionality through parameterised aliases, having it easy to read (user side), cross-platform, and easy to develop and maintain (our side).
* Python script instead of Makefile seems to be flexible and cross-platform. As for the easiness of reading and developing it, I'm not sure.

**Comment by ayushkovskiy** [comment]
> 5. Upgradable (to keep the recipes up-to-date).

User also need their projects to be updated. Currently, we help users to create their custom project that in fact is a mix of configuration+code, so unless we exclude code from there, the users will need to maintain it themselves. Since we assume that the Makefile can be modified by the users in an arbitrary way, it's impossible to provide an automated way to upgrade the Template (only a semi-automated via git).

**Comment by yartem** [comment]
- CI on Windows partially solved cross-platform issues
- so far, aliases offer much less functionality than makefile (no variables, syntax is more complicated than in Makefile)
- we almost completely got rid of posix utilities like `tee`, ... (see `--tag`)

So far, Makefile works fine, closing issue.
